matchtree.go

```diff
@@ -141,6 +141,95 @@ type branchQueryMatchTree struct {
 	docID     uint32
 }
 
+type symbolRegexpMatchTree struct {
+	matchTree
+	regexp *regexp.Regexp
+
+	reEvaluated bool
+	found       []*candidateMatch
+}
+
+func (t *symbolRegexpMatchTree) prepare(doc uint32) {
+	t.reEvaluated = false
+}
+
+func (t *symbolRegexpMatchTree) matches(cp *contentProvider, cost int, known map[matchTree]bool) (bool, bool) {
+	if t.reEvaluated {
+		return len(t.found) > 0, true
+	}
+
+	if cost < costRegexp {
+		return false, false
+	}
+
+	sections := cp.docSections()
+	content := cp.data(false)
+
+	found := t.found[:0]
+	for _, sec := range sections {
+		idx := t.regexp.FindIndex(content[sec.Start:sec.End])
+		if idx == nil {
+			continue
+		}
+		cm := &candidateMatch{
+			byteOffset:  sec.Start,
+			byteMatchSz: uint32(idx[1] - idx[0]),
+		}
+
+		found = append(found, cm)
+	}
+	t.found = found
+	t.reEvaluated = true
+
+	return len(t.found) > 0, true
+}
+
+type symbolSubstrMatchTree struct {
+	*substrMatchTree
+
+	patternSize  uint32
+	fileEndRunes []uint32
+
+	doc      uint32
+	sections []DocumentSection
+}
+
+func (t *symbolSubstrMatchTree) prepare(doc uint32) {
+	t.substrMatchTree.prepare(doc)
+	t.doc = doc
+
+	var fileStart uint32
+	if doc > 0 {
+		fileStart = t.fileEndRunes[doc-1]
+	}
+
+	for len(t.sections) > 0 && t.sections[0].Start < fileStart {
```

Comment:
Zoekt doesn't seem to be doing any skipping to find the starting rune document sections for a file. It's a linear search. Perhaps, we can store additional data (starting rune document section offset per file) in the index file to make this O(1).

Reply:
Another format is to group the sections into batches of 1024 (or some other number). This allows you to skip quickly if your search only applies to a few files. This is something I have been exploring for the posting list. But document jumping also makes sense :+1: